src/parser.rs

```diff
@@ -84,75 +94,40 @@ impl error::Error for Error {
             }
             Error::UnknownAbbreviation => "Found a record with an unknown abbreviation code",
             Error::UnexpectedEof => "Hit the end of input before it was expected",
+            Error::UnknownStandardOpcode(_) => "Found an unknown standard opcode",
+            Error::UnknownExtendedOpcode(_) => "Found an unknown extended opcode",
+            Error::UnsupportedAddressSize(_) => "The specified address size is not supported",
         }
     }
 }
 
 /// The result of a parse.
 pub type ParseResult<T> = Result<T, Error>;
 
-/// A &[u8] slice with compile-time endianity metadata.
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub struct EndianBuf<'input, Endian>(&'input [u8], PhantomData<Endian>) where Endian: Endianity;
-
-impl<'input, Endian> EndianBuf<'input, Endian>
-    where Endian: Endianity
-{
-    pub fn new(buf: &'input [u8]) -> EndianBuf<'input, Endian> {
-        EndianBuf(buf, PhantomData)
-    }
-
-    // Unfortunately, std::ops::Index *must* return a reference, so we can't
-    // implement Index<Range<usize>> to return a new EndianBuf the way we would
-    // like to. Instead, we abandon fancy indexing operators and have these
-    // plain old methods.
-
-    #[allow(dead_code)]
-    fn range_from(&self, idx: RangeFrom<usize>) -> EndianBuf<'input, Endian> {
-        EndianBuf(&self.0[idx], self.1)
-    }
-
-    fn range_to(&self, idx: RangeTo<usize>) -> EndianBuf<'input, Endian> {
-        EndianBuf(&self.0[idx], self.1)
-    }
-}
-
-impl<'input, Endian> Index<usize> for EndianBuf<'input, Endian>
-    where Endian: Endianity
-{
-    type Output = u8;
-    fn index(&self, idx: usize) -> &Self::Output {
-        &self.0[idx]
-    }
-}
-
-impl<'input, Endian> Deref for EndianBuf<'input, Endian>
-    where Endian: Endianity
-{
-    type Target = [u8];
-    fn deref(&self) -> &Self::Target {
-        self.0
-    }
-}
-
-impl<'input, Endian> Into<&'input [u8]> for EndianBuf<'input, Endian>
-    where Endian: Endianity
-{
-    fn into(self) -> &'input [u8] {
-        self.0
+/// Parse a `u8` from the input.
+#[doc(hidden)]
+pub fn parse_u8(input: &[u8]) -> ParseResult<(&[u8], u8)> {
+    if input.len() == 0 {
+        Err(Error::UnexpectedEof)
+    } else {
+        Ok((&input[1..], input[0]))
     }
 }
 
-pub fn parse_u8(input: &[u8]) -> ParseResult<(&[u8], u8)> {
+/// Parse a `i8` from the input.
+#[doc(hidden)]
+pub fn parse_i8(input: &[u8]) -> ParseResult<(&[u8], i8)> {
     if input.len() == 0 {
         Err(Error::UnexpectedEof)
     } else {
-        Ok((&input[1..], input[0]))
+        Ok((&input[1..], unsafe { mem::transmute(input[0]) }))
```

Comment:
Ok I totally misunderstood `as` conversions/coercions -- I thought that `u8 as i8` would always result in either a positive `i8` or a panic if the `u8` was greater than `std::i8::MAX` but actually it does reinterpret the bits directly, just like this transmute intended to do.